File: 74-search-a-2d-matrix/search-a-2d-matrix.java
class Solution
{
    public boolean searchMatrix(int[][] matrix, int target)
    {
        int rows = matrix.length;
        int cols = matrix[0].length;

        int top = 0;
        int bot = rows - 1;
        int row = 0;

        while (top <= bot)
        {
            row = (top + bot) / 2;

            if (matrix[row][0] > target)
            {
                bot = row - 1;
            }
            else if (matrix[row][cols - 1] < target)
            {
                top = row + 1;
            }
            else
            {
                break;
            }
        }

        if (!(top <= bot))
        {
            return false;
        }

        int l = 0;
        int r = cols - 1;

        while (l <= r)
        {
            int mid = l + (r - l) / 2;

            if (matrix[row][mid] > target)
            {
                r = mid - 1;
            }
            else if (matrix[row][mid] < target)
            {
                l = mid + 1;
            }
            else
            {
                return true;
            }
        }
        return false;
    }
}

File: 74-search-a-2d-matrix/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # GETTING ROW AND THEN SEARCHING THE COLUMNS OF THAT ROW 

        Rows , Cols = len(matrix) , len(matrix[0])

        top = 0 
        bot = Rows - 1

        while top <= bot:
            row = (top + bot ) // 2

            if matrix[row][0] > target:
                bot = row - 1
            elif matrix[row][Cols - 1] < target:
                top = row + 1
            else:
                break
        
        if not (top <= bot):
            return False
        
        l = 0 
        r = Cols -1 
        

        while l <= r:
            m = (l + r) // 2

            if matrix[row][m] > target:
                r = m - 1
            elif matrix[row][m] < target:
                l = m + 1
            else:
                return True
        return False



        